landoapi/models/revisions.py

```diff
@@ -32,13 +46,302 @@ class DiffWarningGroup(enum.Enum):
     LINT = "LINT"
 
 
+@enum.unique
+class RevisionStatus(enum.Enum):
+    # New means this revision was just created.
+    NEW = "NEW"
+
+    # Stale means something changed upstream and we need to re-check this revision.
+    STALE = "STALE"
+
+    # Waiting means it can be picked up by the revision worker.
+    WAITING = "WAITING"
+
+    # Picked up means a revision worker has picked this up. This signals to other
+    # workers to not pick up this particular revision. This is really just an
+    # "in between" state.
+    PICKED_UP = "PICKED_UP"
+
+    # Checking means it is currently running through various checks.
+    CHECKING = "CHECKING"
+
+    # Problem means something went wrong in some of the checks.
+    PROBLEM = "PROBLEM"
+
+    # Ready means revision worker is finished and this revision can be queued to land.
+    READY = "READY"
+
+    # Below four statuses describe the landing state.
+    QUEUED = "QUEUED"  # LandingJob has been submitted
+    LANDING = "LANDING"  # LandingWorker is processing job
+    LANDED = "LANDED"  # LandingWorker is finished processing job
+    FAILED = "FAILED"  # LandingWorker could not land job
+
+    @classmethod
+    @property
+    def LANDING_STATES(cls):
+        """States where the revision is in process of landing."""
+        return (cls.QUEUED, cls.LANDING, cls.LANDED)
+
+    @classmethod
+    @property
+    def NON_READY_STATES(cls):
+        return (cls.NEW, cls.STALE, cls.WAITING, cls.CHECKING)
+
+
+class RevisionLandingJob(db.Model):
+    """
+    Keep track of the many-to-many relationship between landing jobs and revisions.
+    """
+
+    landing_job_id = db.Column(db.ForeignKey("landing_job.id"), primary_key=True)
+    revision_id = db.Column(db.ForeignKey("revision.id"), primary_key=True)
+    index = db.Column(db.Integer, nullable=True)
+
+    revision = db.relationship("Revision", back_populates="landing_jobs")
+    landing_job = db.relationship("LandingJob", back_populates="revisions")
+
+
+class Revision(Base):
+    """A Lando revision mapping to a Phabricator revision."""
+
+    PATCH_DIRECTORY = Path("/patches")
```

Comment:
Did we consider using Redis to cache the patches?

Reply:
Interesting idea. I think we should consider moving to `memcached` (for other reasons), and that might be more suitable for caching large files/objects.

Reply:
Redis should be able to handle caching large files/objects. Even the largest of patches I would never expect to exceed a few MB.

Simple patch caching in the container is fine for now, but if we find ourselves putting too much work into the implementation I would say we should put that effort into caching via Redis instead. :)

Reply:
No longer relevant as we are using the new `Revision.patch_bytes` field for this now.